### ISSUE TITLE
ENH: Add option to clip out of bounds input values to minimum and maximum of search array in scipy.interpolate.interp1d()

### ISSUE THREAD

**Issue by EwoutH** [scipy/scipy#16338]
### Is your feature request related to a problem? Please describe.

[scipy.interpolate.interp1d](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp1d.html) currently only allows `“extrapolate”` as method to fill values. If you want to define a tuple with the search bounds, you have to calculate those yourself.

### Describe the solution you'd like.

I would like to propose a new built-in method to fill values, which clips the return values to the minimum and maximum of your search array. This way it can only return values at the bounds of your search array, this would be equivalent to `fill_value=(min(search_array), max(search_array)), bounds_error=False`.

It could be called `"clip"` and used as `fill_value="clip"`.

### Describe alternatives you've considered.

Of course it's not that cumbersome to implement it yourself, but I thought it would be a nice quality-of-life addition. It can also also potentially reach higher performance with a built-in implementation.

**Comment by ev-br** [comment]
TL;DR: I'd rather not add it. 

If this is considered an improvement in user experience, I'm honestly not sure it is. Let's compare the two alternatives you mentioned:
 `interp1d(x, y, fill_value=(x.min(), x.max()), bounds_error=False)`  vs
 `interp1d(x, y, fill_value="clip", bounds_error=False)`

The first form (which works today) is probably as straightforward as it gets. The second one is only marginally shorter keystroke-wise. And then how does it work? I need to RTFM. Then I need to check it because the docs are often stale, so I'm firing up an ipython and type... Precisely these two forms to see that they give the same answer! By now I've typed them both; one is straightforward, the other one is what, does it do something more clever or more tricky or fixes some subtlety I'm not seeing at the moment? I'm now thinking what can go wrong in the `fill_value=(x.min(), x.max())` form. Not seeing it. OK, let's keep simple stuff simple and go with an obvious thing. --- At least that would be my train of thought.
So the net result for a user is the same, only more effort is spent trying to parse the docs and think it through. Not much benefit from a user side.

From the library side, any addition is not a one-liner, it needs tests, docs etc. So not much of a benefit for the library either.

You allude to a performance benefit --- would you mind expanding on how it would/could improve performance?

**Comment by EwoutH** [comment]
Thanks for the extensive response. I tend to agree with your standpoint, but I want to make a few counter points for arguments sake:

> The second one is only marginally shorter keystroke-wise. And then how does it work? I need to RTFM.

This is indeed the case when a variable has a short name. But let's use `descriptive_name` for a medium-sized variable name:

`interp1d(descriptive_name, y, fill_value=(descriptive_name.min(), descriptive_name.max()), bounds_error=False)`
vs
`interp1d(descriptive_name, y, fill_value="clip", bounds_error=False)`

I think `fill_value="clip"` should also imply `bounds_error=False`, just like `fill_value="extrapolate"` currently does. Then it becomes:

`interp1d(descriptive_name, y, fill_value="clip")`

> You allude to a performance benefit --- would you mind expanding on how it would/could improve performance?

Of course. `.min()` and `.max()` are two separate function calls in which the array needs to read out (and sorted?). If `Iscipy.interpolate.interp1d()` does this anyway, there might be a potential performance gain.

Also let's consider the case that `x` is not a single variable, but data structure of some sort, like a dictionary, ndarray or DataFrame. In that case, using `.min()` and `.max()`, the array needed to be called 3 times. This was my original use case. Of course you can define `x = df["column"]` or something, but that's an extra line of code you other wise wont need.

So in my usecase, it became:

```
array = ndarray["column"]
amin, amax = array.min(), array.max()
interp1d(array, y, fill_value=(amin, amax), bounds_error=False)
```
While it could be
```
interp1d(ndarray["column"], y, fill_value="clip")
```

-----

In the end I ended up using [numpy.interp()](https://numpy.org/doc/stable/reference/generated/numpy.interp.html#numpy-interp), because it was faster, and this function was called a lot so performance was important. It also had the default behaviour of clipping the values to the min and max of the array, which was in this case the desired behaviour.

**Comment by ev-br** [comment]
1. Long_descriptive_variable_names_are_nice.  I do believe however that a user who introduced them is thus committing to either typing them or using their IDE auto-completion. 
2. Making kwargs interpedependent --- `fill_value="clip"` implying `bounds_error=False` --- is a big no-no. Let's not add more magic to an already messy situation. If we were to start from scratch, we would not likely add the `bounds_error` kwarg at all (in fact, newer parts of the library do not have it), but as of today, it's frozen by backwards compat. 
3. We can reconsider parts of this proposal if there is a _demonstrated_ performance benefit. So far I fail to see if there's one and I suspect there isn't. Happy to be proven wrong though.

**Comment by ev-br** [comment]
OK, two weeks with no follow-up, and no demonstration of a perf improvement. The observation about `fill_value=(x.min(), x.max())` is nice though, and so is that it's the `np.interp` default. I'm going to mark this suggestion for a mention in the interpolate tutorial, under the _extrapolation tips and tricks_ rubric. Yes, this rubric is yet-to-be written, so if you feel like giving it a start, it'll be much appreciated.
